### DISCUSSION TITLE
How to rebuild widgets without assinging new value to the entire state?

### DISCUSSION THREAD

**Discussion by StephanoGeorge** [rrousselGit/riverpod#1365] [Q&A] [answered]
I have a provider that contains a list of messages, when a message from internet arrives, it needs to be appended at the end of the list, and widgets watching the provider needs rebuilding.
But `state = [...state, message]` is much slower than `state.add(message)`, see benchmark:

```dart
import 'package:benchmark/benchmark.dart';

main() {
  late List<String> list = List.generate(1000, (i) => 'Text: $i');
  group('1', () {
    benchmark('list spread', () {
      [...list, 'last'];
    });
  });

  group('2', () {
    setUpEach(() => list = List.generate(1000, (i) => 'Text: $i'));
    benchmark('list.add', () {
      list.add('last');
    });
  });
}
```

Result:

```
 1
  ✓ list spread (19 us)
 2
  ✓ list.add (4 us)
```

So how to rebuild widgets without assinging new value to entire state? Thanks.

**Comment by rrousselGit** [comment, marked as answer]
StateNotifier uses immutable state

If you do not want immutable state use ChangeNotifier
